bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java

```diff
@@ -68,57 +69,67 @@ protected String getRpcCallbackUrl() {
         if (logger.isTraceEnabled()) {
             logger.trace("Client XmlRpcRequest (port {}):\n{}", port, request);
         }
-        return sendMessage(port, request, 0);
+        Exception reason = null;
+        for (int rpcRetryCounter = 1; rpcRetryCounter <= MAX_RPC_RETRY; rpcRetryCounter++) {
+            try {
+                byte[] response = send(port, request);
+                Object[] data = new XmlRpcResponse(new ByteArrayInputStream(response), config.getEncoding())
+                        .getResponseData();
+                return new RpcResponseParser(request).parse(data);
+            } catch (UnknownRpcFailureException | UnknownParameterSetException ex) {
+                throw ex;
+            } catch (Exception ex) {
+                reason = ex;
+                if ("init".equals(request.getMethodName())) { // no retries for "init" request
+                    break;
+                }
+                logger.debug("XmlRpcMessage failure, sending message again {}/{}", rpcRetryCounter, MAX_RPC_RETRY);
+            }
+        }
+        throw new IOException(reason.getMessage(), reason); // can't be null here because logic in for loop
     }
 
-    /**
-     * Sends the message, retries if there was an error.
-     */
-    private synchronized Object[] sendMessage(int port, RpcRequest<String> request, int rpcRetryCounter)
-            throws IOException {
+    private byte[] send(int port, RpcRequest<String> request) throws Exception {
```

Comment:
Please do not throw `Exception`, either define your own like `HomematicException extends Exception` or use something like `IllegalStateException` (or what fits best) and catch only that one. The problem with catchin `Exception` is that it hides programming errors like NPE.

Reply:
I agree with you regarding the catching and throwing of "Exception", but I tried to stay close to the original implementation and Jetty throws a lot of different exceptions. Will try to get rid of the Exception.